File: base/src/main/java/de/whitefrog/frogr/model/SaveContext.java
package de.whitefrog.frogr.model;

import de.whitefrog.frogr.exception.FrogrException;
import de.whitefrog.frogr.persistence.AnnotationDescriptor;
import de.whitefrog.frogr.persistence.FieldDescriptor;
import de.whitefrog.frogr.persistence.Persistence;
import de.whitefrog.frogr.repository.Repository;
import org.neo4j.graphdb.PropertyContainer;

import java.lang.reflect.Field;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Context for entity/relationship save operations.
 * Takes one entity or relationship and tests for property and relationship changes.
 * @param <T> Entity type
 */
public class SaveContext<T extends Base> {
  /**
   * Model used for this save operation.
   */
  private T model;
  /**
   * The original fetched from database.
   */
  private T original;
  /**
   * The repository used in the save context.
   */
  private Repository<T> repository;
  /**
   * The neo4j node or relationship.
   */
  private PropertyContainer node;
  /**
   * List of FieldDescriptor's containing fields that 
   * changed compared to the original.
   */
  private List<FieldDescriptor> changedFields;
  /**
   * Reference to the model's field map.
   */
  private List<FieldDescriptor> fieldMap;

  public SaveContext(Repository<T> repository, T model) {
    this.repository = repository;
    this.model = model;
    if(model.getId() > -1) {
      original = repository.createModel(node());
    }
    else if(model.getUuid() != null) {
      original = repository.findByUuid(model.getUuid());
      model.setId(original.getId());
    }
    fieldMap = Persistence.cache().fieldMap(model.getClass());
  }

  /**
   * Get a full list of changed fields.
   * @return List of changed fields
   */
  public List<FieldDescriptor> changedFields() {
    if(changedFields == null) {
      if(original() != null) repository.fetch(original(), Entity.AllFields);
      changedFields = fieldMap.stream()
        .filter(f-> fieldChanged(f.field()))
        .collect(Collectors.toList());
    }
    return changedFields;
  }

  /**
   * Test if a single field has changed.
   * @param fieldName Field name.
   * @return True if the field has changed
   */
  public boolean fieldChanged(String fieldName) {
    return changedFields().stream().anyMatch(f -> f.getName().equals(fieldName));
  }
  private boolean fieldChanged(Field field) {
    AnnotationDescriptor annotation = 
      Persistence.cache().fieldAnnotations(repository().getModelClass(), field.getName());
    try {
      if(!field.isAccessible()) field.setAccessible(true);
      Object value = field.get(model);
      if(value != null && !annotation.nullRemove) {
        if(original() == null) {
          return true;
        }
        else {
          if(annotation.relatedTo != null && annotation.lazy) return true;
          if(annotation.relatedTo != null) repository().fetch(original(), FieldList.parseFields(field.getName()+"(max)"));
          Object originalValue = field.get(original());
          return !value.equals(originalValue);
        }
      } else if(annotation.nullRemove) {
        return true;
      }
    } catch(IllegalAccessException e) {
      throw new FrogrException(e.getMessage(), e);
    }
    
    return false;
  }

  /**
   * Get the field map for the model in this context.
   * @return Field map of the model in this context.
   */
  public List<FieldDescriptor> fieldMap() {
    return fieldMap;
  }

  /**
   * Get the current model.
   * @return The model in this context.
   */
  public T model() {
    return model;
  }

  /**
   * Get the model repository.
   * @return The model repository.
   */
  @SuppressWarnings("unchecked")
  public Repository<T> repository() {
    return repository;
  }

  /**
   * Get the neo4j node or relationship used as reference in this context.
   * @return The neo4j node or relationship used as reference in this context.
   */
  @SuppressWarnings("unchecked")
  public <N extends PropertyContainer> N node() {
    if(node == null && original() != null) {
      if(model instanceof Model) node = repository.graph().getNodeById(original().getId());
      else node = repository.graph().getRelationshipById(original().getId());
    } else if(node == null && model().getId() > -1) {
      if(model instanceof Model) node = repository.graph().getNodeById(model().getId());
      else node = repository.graph().getRelationshipById(model().getId());
    }
    return (N) node;
  }

  /**
   * The original model used as reference in this context.
   * @return The original model used as reference in this context.
   */
  public T original() {
    return original;
  }

  /**
   * Set the neo4j node. Used when a new node is created.
   * !!! Should not be used outside of persistence methods.
   * @param node The neo4j node
   */
  public void setNode(PropertyContainer node) {
    this.node = node;
  }
}

File: base/src/main/java/de/whitefrog/frogr/model/SaveContext.kt
package de.whitefrog.frogr.model

import de.whitefrog.frogr.exception.FrogrException
import de.whitefrog.frogr.exception.PersistException
import de.whitefrog.frogr.persistence.FieldDescriptor
import de.whitefrog.frogr.repository.Repository
import org.neo4j.graphdb.PropertyContainer
import java.lang.reflect.Field

/**
 * Context for entity/relationship save operations.
 * Takes one entity or relationship and tests for property and relationship changes.
 */
class SaveContext<T : Base>(
  /**
   * The repository used in the save context.
   */
  private val repository: Repository<T>,
  /**
   * Model used for this save operation.
   */
  private val model: T) {
  /**
   * The original fetched from database.
   */
  private var original: T? = null
  /**
   * The neo4j node or relationship.
   */
  private var node: PropertyContainer? = null
  /**
   * List of FieldDescriptor's containing fields that
   * changed compared to the original.
   */
  private var changedFields: List<FieldDescriptor<*>>? = null
  /**
   * Reference to the model's field map.
   */
  private val fieldMap: List<FieldDescriptor<*>>

  init {
    if (model.id > -1) {
      original = repository.createModel(node())
    } else if (model is FBase && model.uuid != null) {
      original = repository.findByUuid(model.uuid)
      model.id = original!!.id
    }
    fieldMap = repository.cache().fieldMap(model.javaClass)
  }

  /**
   * Get a full list of changed fields.
   * @return list of changed fields
   */
  fun changedFields(): List<FieldDescriptor<*>> {
    if (changedFields == null) {
      if (original() != null) repository.fetch(original(), Base.AllFields)
      changedFields = fieldMap
        .filter { f -> fieldChanged(f.field()) }
    }
    return changedFields!!
  }

  /**
   * Test if a single field has changed.
   * @param fieldName field name
   * @return <code>true</code> if the field has changed, otherwise <code>false</code>
   */
  fun fieldChanged(fieldName: String): Boolean {
    return changedFields().stream().anyMatch { f -> f.name == fieldName }
  }

  private fun fieldChanged(field: Field): Boolean {
    val annotation = repository.cache().fieldAnnotations(repository().modelClass, field.name)
    try {
      if (!field.isAccessible) field.isAccessible = true
      val value = field.get(model)
      if (value != null && !annotation!!.nullRemove) {
        if (original() == null) {
          return true
        } else {
          if (annotation.relatedTo != null && annotation.lazy) return true
          if (annotation.relatedTo != null) repository().fetch(original(), FieldList.parseFields(field.name + "(max)"))
          val originalValue = field.get(original())
          return value != originalValue
        }
      } else if (annotation!!.nullRemove) {
        return true
      }
    } catch (e: IllegalAccessException) {
      throw FrogrException(e.message, e)
    }

    return false
  }

  /**
   * Get the field map for the model in this context.
   * @return field map of the model in this context
   */
  fun fieldMap(): List<FieldDescriptor<*>> {
    return fieldMap
  }

  /**
   * Get the current model.
   * @return the model in this context
   */
  fun model(): T {
    return model
  }

  /**
   * Get the model repository.
   * @return the model repository
   */
  fun repository(): Repository<T> {
    return repository
  }

    /**
   * Get the neo4j node or relationship used as reference in this context.
   * @return the neo4j node or relationship used as reference in this context
   */ 
  @Suppress("UNCHECKED_CAST")
  fun <N : PropertyContainer> node(): N {
    if (node == null) {
      node = if (original() != null) {
        if (model is Model)
          repository.graph().getNodeById(original()!!.id)
        else
          repository.graph().getRelationshipById(original()!!.id)
      } else if (model().id > -1) {
        if (model is Model)
          repository.graph().getNodeById(model().id)
        else
          repository.graph().getRelationshipById(model().id)
      } else {
        throw PersistException("could not resolve model $model")
      }
    }
    return node as N
  }

  /**
   * The original model used as reference in this context.
   * @return the original model used as reference in this context
   */
  fun original(): T? {
    return original
  }

  /**
   * Set the neo4j node. Used when a new node is created.
   * !!! Should not be used outside of persistence methods.
   * @param node the neo4j node
   */
  fun setNode(node: PropertyContainer) {
    this.node = node
  }
}
